src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +32,134 @@ case class NoOpFingerprint() {
 }
 
 // IndexLogEntry-specific Content that uses IndexLogEntry-specific fingerprint.
-case class Content(root: String, directories: Seq[Content.Directory])
+case class Content(root: Directory) {
+
+  // List of fully qualified paths of all files mentioned in this Content object.
+  @JsonIgnore
+  lazy val files: Seq[Path] = {
+    // Recursively append parent dir path to subdirs and files.
+    def rec(prefixPath: Path, directory: Directory): Seq[Path] = {
+      val files = directory.files.map(f => new Path(prefixPath, f.name))
+      files ++ directory.subDirs.flatMap { dir =>
+        rec(new Path(prefixPath, dir.name), dir)
+      }
+    }
+
+    rec(new Path(root.name), root)
+  }
+}
+
 object Content {
-  case class Directory(path: String, files: Seq[FileInfo], fingerprint: NoOpFingerprint)
-  object Directory {
-    // modifiedTime is an Epoch time in milliseconds. (ms since 1970-01-01T00:00:00.000 UTC).
-    case class FileInfo(name: String, size: Long, modifiedTime: Long)
-    object FileInfo {
-      def apply(s: FileStatus): FileInfo =
-        FileInfo(s.getPath.toString, s.getLen, s.getModificationTime)
+  // Create a Content object from a directory path by recursively listing its leaf files. All
+  // files from the directory tree will be part of the Content.
+  def apply(path: Path): Content = Content(Directory(path))
+
+  // Create a Content object from a specified list of leaf files. Any files not listed here will
+  // NOT be part of the returned object
+  def fromLeafFiles(files: Seq[FileStatus]): Content = {
+    /* from org.apache.spark.sql.execution.datasources.PartitionAwareFileIndex. */
+    val leafDirToChildrenFiles: Map[Path, Array[FileStatus]] =
+      files.toArray.groupBy(_.getPath.getParent)
+    val rootPath = getRoot(leafDirToChildrenFiles.head._1)
+
+    val subDirs = leafDirToChildrenFiles
+      .filterNot(_._1.isRoot)
+      .map {
+        case (dir, statuses) => createSubDirectoryTree(dir, statuses)
+      }
+      .toSeq
+
+    val rootFiles: Seq[FileInfo] =
+      leafDirToChildrenFiles.getOrElse(rootPath, Array()).map(FileInfo(_))
+
+    Content(Directory(rootPath.toString, rootFiles, subDirs))
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
```

Comment:
`Path` also has `getRoot`. I guess that's not the behavior you wanted? If not, could you add a comment?

Reply:
I don't think path has `getRoot`, could you please reconfirm or correct me if I am wrong? [github](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java)
Is it by some other name?

Reply:
closing it for now, please feel free to reopen